filemapping.go

```diff
@@ -0,0 +1,54 @@
+package gettext
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"runtime"
+)
+
+type fileMapping struct {
+	data []byte
+
+	isMapped bool
+}
+
+func (m *fileMapping) Close() error {
+	runtime.SetFinalizer(m, nil)
+	if !m.isMapped {
+		return nil
+	}
+	return m.closeMapping()
+}
+
+func openMapping(f *os.File) (*fileMapping, error) {
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	m := new(fileMapping)
+
+	if fi.Mode().IsRegular() {
+		size := fi.Size()
+		if size == 0 {
+			return m, nil
+		}
+		if size < 0 {
```

Comment:
Should we store size in the "fileMapping" so that we can do out-of-bounds checks? My worry is that a buggy mo file could crash snapd easily if the indexes point to regions out of file and we would not detect this but crash instead, is this so or am I missing some other safeguard?

Reply:
The size of the memory mapped region is stored in the length of the `data` slice.  So access to the memory region is subject to the same bounds checks as any other slice (i.e. out of bounds access will panic).

As the existing library API does not have a way to report errors for each translation, I am performing some validation on the catalogue at the time of loading:

1. That the original strings table is within bounds (by checking that the `OrigTabOffset + 8*NumStrings` is not past the end of the file).
2. The same for the translated strings table.
3. For both of the above tables, that the strings they represent are within bounds.
4. If the catalogue contains a hash table that every entry is less than or equal to `NumStrings + 1`.  Zeroes represent empty buckets, otherwise they are a string index incremented by one.

As we map the file with MAP_PRIVATE, we don't need to worry about the data changing underneath us either, so we shouldn't trigger any out of bounds reads while looking up translations.